apps/meteor/app/models/server/raw/Sessions.ts

```diff
@@ -745,17 +747,313 @@ export class SessionsRaw extends BaseRaw<ISession> {
 		this.secondaryCollection = colSecondary;
 	}
 
+	async getByUserId(
+		uid: string,
+		search?: string | null,
+		{ offset, count }: IPaginationOptions = { offset: 0, count: 50 },
+	): Promise<PaginatedResult<{ sessions: ISession[] }>> {
+		const searchQuery = search ? [{ $text: { $search: search } }] : [];
+
+		const matchOperator = {
+			$match: {
+				$and: [
+					...searchQuery,
+					{
+						userId: {
+							$eq: uid,
+						},
+					},
+					{
+						loginToken: {
+							$exists: true,
+							$ne: '',
+						},
+					},
+					{
+						logoutAt: {
+							$exists: false,
+						},
+					},
+				],
+			},
+		};
+		const sortOperator = {
+			$sort: {
+				createdAt: -1,
+			},
+		};
+		const groupOperator = {
+			$group: {
+				_id: '$loginToken',
+				instanceId: {
+					$first: '$instanceId',
+				},
+				sessionId: {
+					$first: '$sessionId',
+				},
+				day: {
+					$first: '$day',
+				},
+				month: {
+					$first: '$month',
+				},
+				year: {
+					$first: '$year',
+				},
+				userId: {
+					$first: '$userId',
+				},
+				device: {
+					$first: '$device',
+				},
+				host: {
+					$first: '$host',
+				},
+				ip: {
+					$first: '$ip',
+				},
+				mostImportantRole: {
+					$first: '$mostImportantRole',
+				},
+				roles: {
+					$first: '$roles',
+				},
+				type: {
+					$first: '$type',
+				},
+				_updatedAt: {
+					$first: '$_updatedAt',
+				},
+				createdAt: {
+					$first: '$createdAt',
+				},
+				loginAt: {
+					$first: '$loginAt',
+				},
+			},
+		};
+
+		const skipOperator = offset >= 1 ? [{ $skip: offset }] : [];
+		const limitOperator = { $limit: count };
+
+		const projectOperator = {
+			$project: {
+				_id: '$sessionId',
+				instanceId: 1,
+				sessionId: 1,
+				day: 1,
+				month: 1,
+				year: 1,
+				userId: 1,
+				device: 1,
+				host: 1,
+				ip: 1,
+				mostImportantRole: 1,
+				roles: 1,
+				type: 1,
+				_updatedAt: 1,
+				createdAt: 1,
+				loginAt: 1,
+			},
+		};
+
+		const facetOperator = {
+			$facet: {
+				docs: [sortOperator, ...skipOperator, limitOperator],
+				count: [
+					{
+						$count: 'total',
+					},
+				],
+			},
+		};
+
+		const queryArray = [matchOperator, sortOperator, groupOperator, projectOperator, facetOperator];
+
+		const [
+			{
+				docs: sessions,
+				count: [{ total } = { total: 0 }],
+			},
+		] = (await this.col.aggregate(queryArray).toArray()) as unknown as { docs: ISession[]; count: { total: number }[] }[];
```

Comment:
you can pass a type param to the aggregate function iirc

Reply:
Done